### ISSUE TITLE
Make [unit_placed] no undoable by default

### ISSUE THREAD

**Issue by Toranks** [wesnoth/wesnoth#7780]
### Describe the desired feature

[unit_placed] allows undo, and it's giving me a headache trying to get it to avoid that.

The use of the undocumented function wesnoth.allow_undo(false) is not working right every time, and I can't reproduce the behavior.

It would be preferable to change this behavior and use [allow_undo] where desired.

Note: The only 100% reliable way I know of at the moment to avoid the undo is this code inside the event
```
    #Make the event not undoable
    {VARIABLE_OP disallow_undo rand(0..1)}
    {CLEAR_VARIABLE disallow_undo}
```

**Comment by Wedge009** [comment]
If `wesnoth.allow_undo(false)` is not working as expected, isn't that also a bug? Undocumented or not.

**Comment by Toranks** [comment]
> If `wesnoth.allow_undo(false)` is not working as expected, isn't that also a bug? Undocumented functionality or not.

Yes probably. But since it's not documented I don't know if I'm doing it right or not, so I tagged the only person I've seen using it by googling/search forum.

**Comment by Toranks** [comment]
If it helps, the two ways I've tried to get it to work are these:
```
         function wesnoth.wml_actions.deny_undo()
             wesnoth.allow_undo(false)
         end
```
And using [deny_undo] the same way as [allow_undo], without keys


I also tried this directly in the unit placed event:
```
        [lua]
            code=<<
                wesnoth.allow_undo(false)
            >>
        [/lua]
```

**Comment by CelticMinstrel** [comment]
Related: #7566, which makes `wesnoth.allow_undo(false)` official API (under a slightly different name) and adds `[disallow_undo]`.

**Comment by gfgtdf** [comment]
> The only example is from @gfgtdf here:

I think if the concrete event (by that i mean the caller on the c++ side) itself lacks the code to prevent undoing then `wesnoth.allow_undo(false)` this doesn't mean that ` wesnoth.allow_undo(false)`is generally broken. That said i still don't the was the api works is  optimal in that regard.

**Comment by gfgtdf** [comment]
It would be rather easy to have a api function/wmltag lets call it for example `[deny_undo]` that would strictly block undoing in a way that it also cannot be re-allowed (just call `synced_context::block_undo()`). `wesnoth.allow_undo(false)` sets the current events  "undoable" flag which can be changed later (epsecially if we consider umc events that fire other events.)

**Comment by CelticMinstrel** [comment]
I think making it non-reversible might be a step too far…?

**Comment by ProditorMagnus** [comment]
The way I would expect events and undo work is that any event which is not with [allow_undo] is counted as disabling undo. And any action with at least one event that disables undo will not have undo.

When action fired multiple event then they all need to have [allow_undo] to provide undo support.

**Comment by gfgtdf** [comment]
> When action fired multiple event then they all need to have [allow_undo] to provide undo support.

This is iirc not the way it currently works, currently if an event A with [allow_undo] fires another event B that doesn't use [allow_undo], then whether the action is undoable depends on what A does first: firing the other event or using [allow_undo]

**Comment by ProditorMagnus** [comment]
I mean
```
[event]
	name=moveto
	first_time_only=no
[/event]
[event]
	name=moveto
	first_time_only=no
	[allow_undo]
	[/allow_undo]
[/event]
```
I expect disable undo. For manually done [fire_event] would be nice to follow same rules but not the main use case.

**Comment by ProditorMagnus** [comment]
Undo is not synced action, so not in [do_command]. Is there way to unit test those?

**Comment by gfgtdf** [comment]
> Undo is not synced action, so not in [do_command]. Is there way to unit test those?

Afaik not

**Comment by CelticMinstrel** [comment]
We should probably add something like `wesnoth.interface.undo` to the API, for unit tests if nothing else.

**Comment by gfgtdf** [comment]
> We should probably add something like wesnoth.interface.undo to the API, for unit tests if nothing else.

What would that do exactly? Just query whether something is undoable? And would it include the currently executed action or not?

**Comment by CelticMinstrel** [comment]
No, it would execute the undo.

**Comment by ProditorMagnus** [comment]
In order to execute undo there needs to be another API to do action which is not immediately synced.

**Comment by gfgtdf** [comment]
The tests actions are never undoable though since technically all their code happens in no undoable events like "side turn", even if they end up doing attacks those attacks are just executed WML and not their own independent actions.


Also I think undoing an action during a other actions is by design quite difficult.
